BUG: Slicing subclasses of SparseDataFrames.

sstanovnik opened the pull request:
- [X] 1 test added & passed
- [X] passes `git diff upstream/master | flake8 --diff`

This changes `SparseDataFrame` to use proper subclassing functionality so slicing of subclasses of `SparseDataFrame` works. Example of a failure that this PR fixes:

``` python
import pandas as pd

class DenseSubclassDF(pd.DataFrame):
    _constructor = property(lambda self: DenseSubclassDF)
    _constructor_sliced = property(lambda self: DenseSubclassS)

class DenseSubclassS(pd.Series):
    _constructor = property(lambda self: DenseSubclassS)
    _constructor_expanddim = property(lambda self: DenseSubclassDF)

class SparseSubclassDF(pd.SparseDataFrame):
    _constructor = property(lambda self: SparseSubclassDF)
    _constructor_sliced = property(lambda self: SparseSubclassS)

class SparseSubclassS(pd.SparseSeries):
    _constructor = property(lambda self: SparseSubclassS)
    _constructor_expanddim = property(lambda self: SparseSubclassDF)

ddf = DenseSubclassDF([[1,2,3], [4,5,6], [7,8,9]])
sdf = SparseSubclassDF([[1,2,3], [4,5,6], [7,8,9]])

print(type(ddf.iloc[0]))  # <class '__main__.DenseSubclassS'>
print(type(ddf.iloc[:2]))  # <class '__main__.DenseSubclassDF'>
print(type(ddf[:2]))  # <class '__main__.DenseSubclassDF'>

# sparse doesn't preserve types
print(type(sdf.iloc[0]))  # <class '__main__.SparseSubclassS'>
print(type(sdf.iloc[:2]))  # <class 'pandas.sparse.frame.SparseDataFrame'>
print(type(sdf[:2]))  # <class 'pandas.sparse.frame.SparseDataFrame'>
```

sstanovnik committed:
BUG: Fix slicing subclasses of SparseDataFrames.

Use proper subclassing behaviour so subclasses work properly: this fixes
an issue where a multi-element slice of a subclass of SparseDataFrame
returned the SparseDataFrame type instead of the subclass type.

jreback commented on pandas/sparse/frame.py:
```diff
@@ -373,7 +373,9 @@ def _slice(self, slobj, axis=0, kind=None):
             new_columns = self.columns[slobj]
 
-        return self.reindex(index=new_index, columns=new_columns)
+        return self._constructor(data=self.reindex(index=new_index,
```
try not to use ``\``, instead:

```
return self._constructor(
          data=self.reindex(index=new_index, 
                                        columns=new_columns)).__finalize__(self)
```

sinhrks replied:
can't we fix ``.reindex`` to return subclass?

jreback commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -210,3 +210,17 @@ def test_subclass_align_combinations(self):
         tm.assert_frame_equal(res2, exp1)
+
+    def test_subclass_sparse_slice(self):
+        ssdf = tm.SubclassedSparseDataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
```
add the issue number here

sstanovnik replied:
I haven't created an issue for this. Should I?

jreback replied:
no, but have a look thru existing ones to see if it covers anything, it might

jreback replied:
just search, maybe sparse (there are not too many open ones)

jreback commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -210,3 +210,17 @@ def test_subclass_align_combinations(self):
+        ssdf = tm.SubclassedSparseDataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+        ssdf.testattr = "testattr"
+
+        tm.assertIsInstance(ssdf.loc[:2], tm.SubclassedSparseDataFrame)
```
use ``tm.assert_sp_frame_equal`` here (and just use a constructed ``SubClassedSparseDataFrame`` as the expected

sstanovnik replied:
As far as I can tell, `tm.assert_sp_frame_equal` only asserts both are instances of `SparseDataFrame` and not that they are the same type. What I meant to fix and test is only preserving the types when slicing. Should I add (not replace with) an additional check with `tm.assert_sp_frame_equal` anyway?

jreback replied:
add a ``check_dtype`` option in testing function which will assert that not only they are instances of ``pd.SparseDataFrame`` , but also that they are the *same* (e.g. signature could be
``check_dtype=True|'same'``

sinhrks replied:
``assert_frame_equal`` and ``assert_series_equal`` have ``check_frame/series_type`` for the same meaning. Can use same options.

jreback commented:
is there a specific issue for this?

@sinhrks pls have a look

sstanovnik committed:
Formatting changes.

sstanovnik committed:
Add type checks to sparse object comparisons.

sstanovnik committed:
Modify the new test to compare complete objects.

sstanovnik committed:
Use _constructor when instantiating SparseDataFrames.

This should fix subclassing issues.

sstanovnik commented:
Thank you for your comments, I made changes according to them:
 - `.reindex` didn't need to be adapted, the problem was `_reindex_columns` which explicitly instantiated a `SparseDataFrame`. I changed all such calls to use `_constructor`.
 - The thing to check in `testing.py` is that I added `obj` as a kwarg to panel checks to conform to what the series and frame ones look like. Previously, `obj` was referenced in them but the one that was used was a leftover from the for loop around line 2600. 
 - There is no issue that handles this problem, the closest one is the sparse master issue #10627. Reference that one or create a new one?

sinhrks commented on pandas/util/testing.py:
```diff
@@ -1422,14 +1446,29 @@ def assert_sp_series_equal(left, right, exact_indices=True,
+
+    Parameters
+    ----------
+    left : DataFrame
```
nice docstring:) `SparseDataFrame`.

sinhrks commented on pandas/util/testing.py:
```diff
@@ -1453,10 +1492,29 @@ def assert_sp_frame_equal(left, right, exact_indices=True,
 
 
-def assert_sp_panel_equal(left, right, exact_indices=True):
+def assert_sp_panel_equal(left, right, exact_indices=True,
```
`SparsePanel` has been removed, needs rebase (#13778).

sinhrks commented:
@sstanovnik thx for update! i linked the PR from #10627, so new issue is not needed unless you find new issue which is not fixid in this PR.

sstanovnik commented:
Should I also squash the commits?

sinhrks commented:
not mandatory on your side. will be done during merge.

sstanovnik commented:
Typo fixed, rebased.

sstanovnik committed:
Lint corrections.

sstanovnik committed:
Further SparseSeries subclass fixes.

sstanovnik commented:
Two old pickling tests failed: see [this build](https://travis-ci.org/pydata/pandas/jobs/147445146#L1263). I didn't see a nice way out, so I made an exception for the deprecated SparseTimeSeries. This may not be what you want and needs review.

sinhrks commented:
@sstanovnik yeah ``SparseTimeSeries`` is deprecated, you can switch validation option depending on the version, like:

- https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_pickle.py#L120

jreback commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -210,3 +210,25 @@ def test_subclass_align_combinations(self):
         tm.assertIsInstance(res2, tm.SubclassedDataFrame)
         tm.assert_frame_equal(res2, exp1)
+
+    def test_subclass_sparse_slice(self):
```
we have lots of changes, but want to see if we can have a test for each, can you audit these changes and add tests if needed.

sstanovnik committed:
Proper test parameter override, travis soothing.

sstanovnik committed:
Additional sparse subclass tests.

sstanovnik commented:
I added additional tests to check other subclassing changes I made.

jreback commented:
lgtm; though I think the read_pickle assertion can be done in a clearer way (use sep function). @sinhrks ?

sinhrks commented on pandas/io/tests/test_pickle.py:
```diff
@@ -44,8 +44,15 @@ def compare_element(self, result, expected, typ, version=None):
 
         if typ.startswith('sp_'):
+            # SparseTimeSeries deprecated in 0.17.0
+            if (typ == "sp_series" and version and
```
``test_pickle`` can use object-specific method if it is defined. Defining ``compare_sp_series_ts`` (like other ``compare_xxx`` method) should allow ``test_pickle`` to use the method for sparse time series. Current impl skips type check of all sparse series.

NOTE: method name is derived from pickle key:

- https://github.com/pydata/pandas/blob/master/pandas/io/tests/generate_legacy_storage_files.py#L157

sstanovnik committed:
Properer test assertion override.

sinhrks commented on pandas/io/tests/test_pickle.py:
```diff
@@ -86,6 +86,13 @@ def compare(self, vf, version):
 
+    def compare_sp_series_ts(self, res, exp, typ, version):
+        # SparseTimeSeries deprecated in 0.17.0
+        if version and LooseVersion(version) < "0.17.0":
```
Thx for the update. I think `` ... <= "0.12.0"`` is OK based on the error below. After that, ``SparseTimeSeries`` has been integrated to ``SparseSeries``.

- https://travis-ci.org/pydata/pandas/jobs/147445144

sstanovnik committed:
Changed the override from 0.17.0 to 0.12.0.

jreback commented:
@sstanovnik looks good. @sinhrks anything further?

sinhrks commented:
None. thanks for all the effort, @sstanovnik !

jreback commented:
@sstanovnik lgtm. pls add a whatsnew note (put in API changes). ping when green.

sstanovnik committed:
Add a whatsnew note.

sstanovnik commented:
ping

jreback commented:
thanks @sstanovnik
